### DISCUSSION TITLE
defaultValues that contains a date is not rendering in input

### DISCUSSION THREAD

**Discussion by kiwiinlondon** [react-hook-form/react-hook-form#7162] [Q&A] [answered]
I am quite new to react and javascript so please excuse me if this is a bit of a rudimentary question or whether in fact this is a bug. 

The simplified code is below or can be seen at https://codesandbox.io/s/react-form-date-issue-327fh?file=/src/App.tsx:48-63
```
import React from 'react';
import ReactDOM from 'react-dom';
import { useForm } from "react-hook-form";

type Person = {
  firstname :string;
  lastname :string;
  dob : Date;
  }

const App = (person: Person) => {

  const { register, handleSubmit } = useForm<Person>( {defaultValues: person});
  
  const onSubmit = handleSubmit((data) => {
    alert(JSON.stringify(data))
  })

  return (    
    <form onSubmit={onSubmit}>
      <input {...register("firstname", {required : 'Required'})} id = "firstname" type="text"/>
      <input {...register("lastname", {required : 'Required'})} id = "lastname" type="text"/>
      <input {...register("dob", {required : 'Required'})} id = "dob" type="date"/>      
      <input type="submit" />
    </form>
  );
}

ReactDOM.render(
  <React.StrictMode>
    <App firstname = "Joe" lastname = "Blogs" dob = {new Date(2002,12,31)} />
  </React.StrictMode>,
  document.getElementById('root')
);
```
This gives me the following form which is prepopulated except for the dd/mm/yyyy enter image description here
![image](https://user-images.githubusercontent.com/95145947/143719749-0ac0deeb-b257-4380-9b70-7a0be6148576.png)

The problem I have is

I have a simple react hook form that uses a type which is used by useForm. In this case its called Person
This form will be populated with defaultvalues that are retrieved from an API.
The type that this form uses has a date in it and I have no control over the object passed in.
When the date is set as part of the defaultvalues in the useForm it is not rendered and all I get is dd/mm/yyyy
When I submit the form the correct value is submitted
My question is how do I get the dob date box to display the underlying date rather than dd/mm/yyyy?

I can change the dob input to be of type text rather than date and the correct value is displayed but then there is of course no validation of the date or a date picker

Rather than passing in the type to defaultValues of useform I have also tried passing in a defaultValue to each of the inputs. This renders the date correctly (if i output the date to a string of correct format) but I would rather not have to set the defaultValue on each input as the type I am actually using is quite large and I thought it was quite nice just to bind the one object
```
<input {...register("dob", {required : 'Required'})} id = "lastname" type="date" defaultValue = {person.dob.toISOString().substr(0, 10)}/>  
```
I know that a input of date requires the defaultvalue to be a text value. However, if I set the default value using defaultvalues in useForm I have no control over how the Input is initialised

 Any help would be greatly appreciated

**Comment by fahadsohail482** [comment, marked as answer]
Method 1:
@kiwiinlondon you may want to use the `reset()` function from `useform()` on mount and set your values as you want. This way you have full control over the values you set to the form.
Here is the snippet on how you can do it.
![image](https://user-images.githubusercontent.com/46647496/143831724-722953ff-215e-4207-9a4a-e36021a472a1.png)

Working CodeSandBox [here](https://codesandbox.io/s/react-form-date-issue-forked-rpi0q?file=/src/App.tsx).

Method 2:
you can use `setValue()` function the similar way but its for the single value use `setValue("name","value")`.

I would prefer you the 1st method. Hoping that it would help you.

**Reply by tan00001** [reply]
This does not work in TypeScript. You cannot assign a string to a Date.

**Reply by RobJJ** [reply]
I've run into a similar problem and would rather not convert the type to a string - this would then affect my zod validation. Anyone have a different solution for this problem?

**Comment by kiwiinlondon** [comment]
Thanks for your suggestion @fahadsohail482. However, to do that I would still have to change the type of the dob from Date to any. What if I dont have control over the Person object?

**Reply by fahadsohail482** [reply]
@kiwiinlondon You will always have control over the props or the data you are fetching from remote. You can change/modify or format as per your liking. 
You don't need to use any type.
![image](https://user-images.githubusercontent.com/46647496/143856134-fadaca52-56e2-495e-849c-83fd22d2ab49.png)
**This component is expecting a `Date` but your `useForm()` is expecting a string see updated [code sandbox](https://codesandbox.io/s/react-form-date-issue-forked-x5h99?file=/src/App.tsx)**

**Reply by tan00001** [reply]
Blame it on string types is not the answer. If the variable is a Date, convert it to a Date. It is not rock science. Why all the excuse for not handling as a Date? You already have an option "valueAsDate", yet when it comes in as a Date, you don't display it. That is just lame.

**Comment by kiwiinlondon** [comment]
Thanks @fahadsohail482. I can work with that.

Excuse me if this is a basic question but is there a reason that react-hook-form could not not take care of this in the background? 
I  am just just starting with React and Javascript, having come from a strongly typed lanaguage background, so would have thought if I had a date on an object and was rendering that object the date would just work

**Reply by fahadsohail482** [reply]
@kiwiinlondon The library does not genuinely anticipate what the user desires. So it provide the utility functions to do stuff.
For Date it has the special case you can read [here.](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/date#:~:text=Note%3A%20The%20displayed%20date%20format%20will%20differ%20from%20the%20actual%20value%20%E2%80%94%20the%20displayed%20date%20is%20formatted%20based%20on%20the%20locale%20of%20the%20user%27s%20browser%2C%20but%20the%20parsed%20value%20is%20always%20formatted%20yyyy%2Dmm%2Ddd.)

**Comment by kiwiinlondon** [comment]
Thanks @fahadsohail482. I suspected that was the answer
